File: neb_pathmaker_core.mjs
import {
  LANGUAGE_OPTIONS,
  NEB_DEFAULTS,
  NEB_SINGLE_HIDDEN_TASKS,
  NEB_TASKS,
  slugify,
} from "./neb_core.mjs?v=2026-07-10-spanish-sub";

const PATHMAKER_BUCKET = "gacm-axinom-staging";
const FIELD_SKU = "sku";
const PATHMAKER_SERIES_TASK = "Series";

const PATHMAKER_DEFAULTS = {
  ...NEB_DEFAULTS,
  [FIELD_SKU]: "",
};
const PATHMAKER_EXTRA_USAGE_TO_PREFIX = {
  "Behind the Scenes / Making Of": "bts",
  "Interviews (Cast/Crew)": "int",
  "Deleted Scenes": "del",
  "Bloopers / Alternate Takes": "alt",
  "Promotional Clips": "clp",
};

const PATHMAKER_TASKS = {};
Object.entries(NEB_TASKS).forEach(([task, definition]) => {
  PATHMAKER_TASKS[task] = {
    ...definition,
    fields: [...definition.fields, FIELD_SKU],
    pathFields: definition.fields.filter((field) => !["resolution", "house", "subtitle_type"].includes(field)).concat(FIELD_SKU),
  };
  if (task === "Movie") {
    PATHMAKER_TASKS[PATHMAKER_SERIES_TASK] = {
      fields: ["title", "language", FIELD_SKU],
      pathFields: ["title", "language", FIELD_SKU],
      pathOnly: true,
    };
  }
});

const MOVIE_ROOT_TASKS = new Set([
  "Movie",
  "Caption",
  "Dub Audio",
  "Virtual Screening",
  "Trailer",
  "Trailer Caption",
  "Extras",
]);

const SERIES_ROOT_TASKS = new Set([
  PATHMAKER_SERIES_TASK,
  "Episode",
  "Episode Caption",
  "Original Premium Series (Yearly)",
  "Exclusive Conversation (Yearly)",
  "Virtual Screening Episode",
  "Virtual Screening Episode Caption",
]);

function normalizePathLanguage(value) {
  const language = String(value || "").trim();
  if (!LANGUAGE_OPTIONS.includes(language)) {
    throw new Error("Language must be English or Spanish.");
  }
  return language;
}

function normalizeSku(value) {
  const sku = String(value || "").trim().toLowerCase();
  if (!sku) {
    throw new Error("SKU / Parent SKU is required.");
  }
  if (!/^[a-z0-9][a-z0-9-]*$/.test(sku)) {
    throw new Error("SKU / Parent SKU can only use letters, numbers, and hyphens.");
  }
  return sku;
}

function normalizeSeasonFolder(value) {
  const raw = String(value || "").trim().toLowerCase().replace(/^s/, "");
  if (!/^\d+$/.test(raw)) {
    throw new Error("Season must be a number like 02.");
  }
  return `s${raw.padStart(2, "0")}`;
}

function normalizeYearSeasonFolder(value) {
  const raw = String(value || "").trim().toLowerCase().replace(/^s/, "");
  if (!/^\d{4}$/.test(raw)) {
    throw new Error("Year must be four digits.");
  }
  return `s${raw}`;
}

function normalizeEpisodeFolder(value) {
  const raw = String(value || "").trim().toLowerCase().replace(/^e/, "");
  if (!/^\d+$/.test(raw)) {
    throw new Error("Episode must be a number like 08.");
  }
  return `e${raw.padStart(2, "0")}`;
}

function titleSlug(rawFields, label = "Title") {
  const slug = slugify(rawFields.title || "");
  if (!slug) {
    throw new Error(`${label} is required.`);
  }
  return slug;
}

function localizedFeatureFolder(language) {
  return language === "Spanish" ? "feature_las" : "feature";
}

function localizedSeasonFolder(seasonFolder, language) {
  return language === "Spanish" ? `${seasonFolder}_las` : seasonFolder;
}

function projectFolder(root, slug, sku) {
  return `s3://${PATHMAKER_BUCKET}/${root}/${slug}_${sku}`;
}

function extraFolder(rawFields) {
  const usage = String(rawFields.extra_usage || "").trim();
  const prefix = PATHMAKER_EXTRA_USAGE_TO_PREFIX[usage];
  if (!prefix) {
    throw new Error("Choose an Extra Type supported by the source file structure doc.");
  }
  const usageSlug = slugify(usage);
  if (!usageSlug) {
    throw new Error("Extra Type is required.");
  }
  return `${prefix}_${usageSlug}`;
}

function rootForTask(task) {
  if (MOVIE_ROOT_TASKS.has(task)) return "movies";
  if (SERIES_ROOT_TASKS.has(task)) return "series";
  throw new Error("Unsupported PathMaker task type.");
}

function projectSlugForTask(task, rawFields) {
  if (task === "Exclusive Conversation (Yearly)") {
    return "exclusive_conversation";
  }
  return titleSlug(rawFields, task === PATHMAKER_SERIES_TASK || task.includes("Episode") ? "Series Title" : "Title");
}

function pathmakerFieldsForTask(task, { forPathOnly = false } = {}) {
  const definition = PATHMAKER_TASKS[task];
  if (!definition) {
    throw new Error("Unsupported PathMaker task type.");
  }
  return forPathOnly ? definition.pathFields : definition.fields;
}

function buildPathmakerPath(task, rawFields) {
  if (!PATHMAKER_TASKS[task]) {
    throw new Error("Unsupported PathMaker task type.");
  }

  const language = normalizePathLanguage(rawFields.language || PATHMAKER_DEFAULTS.language);
  const sku = normalizeSku(rawFields[FIELD_SKU]);
  const root = rootForTask(task);
  const projectSlug = projectSlugForTask(task, rawFields);
  const base = projectFolder(root, projectSlug, sku);

  if (task === PATHMAKER_SERIES_TASK) {
    return `${base}/`;
  }

  if (task === "Movie" || task === "Caption" || task === "Dub Audio") {
    return `${base}/${localizedFeatureFolder(language)}/`;
  }

  if (task === "Episode" || task === "Episode Caption") {
    const season = localizedSeasonFolder(normalizeSeasonFolder(rawFields.season), language);
    const episode = normalizeEpisodeFolder(rawFields.episode);
    return `${base}/${season}/${episode}/`;
  }

  if (task === "Original Premium Series (Yearly)" || task === "Exclusive Conversation (Yearly)") {
    const season = localizedSeasonFolder(normalizeYearSeasonFolder(rawFields.year), language);
    const episode = normalizeEpisodeFolder(rawFields.episode);
    return `${base}/${season}/${episode}/`;
  }

  if (task === "Virtual Screening") {
    return `${base}/feature_virtual_screening/`;
  }

  if (task === "Virtual Screening Episode" || task === "Virtual Screening Episode Caption") {
    const season = localizedSeasonFolder(normalizeSeasonFolder(rawFields.season), language);
    const episode = normalizeEpisodeFolder(rawFields.episode);
    return `${base}/${season}/${episode}_virtual_screening/`;
  }

  if (task === "Trailer" || task === "Trailer Caption") {
    return `${base}/trailer/`;
  }

  if (task === "Extras") {
    return `${base}/extras/${extraFolder(rawFields)}/`;
  }

  throw new Error("Unsupported PathMaker task type.");
}

export {
  FIELD_SKU,
  PATHMAKER_BUCKET,
  PATHMAKER_DEFAULTS,
  PATHMAKER_EXTRA_USAGE_TO_PREFIX,
  PATHMAKER_SERIES_TASK,
  PATHMAKER_TASKS,
  NEB_SINGLE_HIDDEN_TASKS as PATHMAKER_SINGLE_HIDDEN_TASKS,
  buildPathmakerPath,
  pathmakerFieldsForTask,
};

File: neb_pathmaker_core.py
"""S3 path generation for NEB+PathMaker."""

from __future__ import annotations

import re

from axinom_name_builder import (
    DEFAULT_VALUES,
    FIELD_EPISODE,
    FIELD_EXTRA_USAGE,
    FIELD_LANGUAGE,
    FIELD_SEASON,
    FIELD_TITLE,
    FIELD_YEAR,
    LANGUAGE_OPTIONS,
    SINGLE_HIDDEN_TASKS,
    TASKS,
    slugify,
)

PATHMAKER_BUCKET = "gacm-axinom-staging"
FIELD_SKU = "sku"
PATHMAKER_SERIES_TASK = "Series"
PATHMAKER_DEFAULTS = {
    **DEFAULT_VALUES,
    FIELD_SKU: "",
}

PATHMAKER_EXTRA_USAGE_TO_PREFIX = {
    "Behind the Scenes / Making Of": "bts",
    "Interviews (Cast/Crew)": "int",
    "Deleted Scenes": "del",
    "Bloopers / Alternate Takes": "alt",
    "Promotional Clips": "clp",
}

PATHMAKER_TASK_FIELDS: dict[str, tuple[str, ...]] = {}
for task_name, task_definition in TASKS.items():
    PATHMAKER_TASK_FIELDS[task_name] = tuple(task_definition["fields"])  # type: ignore[index]
    if task_name == "Movie":
        PATHMAKER_TASK_FIELDS[PATHMAKER_SERIES_TASK] = (FIELD_TITLE, FIELD_LANGUAGE)

MOVIE_ROOT_TASKS = {
    "Movie",
    "Caption",
    "Dub Audio",
    "Virtual Screening",
    "Trailer",
    "Trailer Caption",
    "Extras",
}

SERIES_ROOT_TASKS = {
    PATHMAKER_SERIES_TASK,
    "Episode",
    "Episode Caption",
    "Original Premium Series (Yearly)",
    "Exclusive Conversation (Yearly)",
    "Virtual Screening Episode",
    "Virtual Screening Episode Caption",
}


def pathmaker_task_options() -> tuple[str, ...]:
    return tuple(task for task in PATHMAKER_TASK_FIELDS if task not in SINGLE_HIDDEN_TASKS)


def pathmaker_fields_for_task(task: str, *, for_path_only: bool = False) -> tuple[str, ...]:
    if task not in PATHMAKER_TASK_FIELDS:
        raise ValueError("Unsupported PathMaker task type.")
    fields = PATHMAKER_TASK_FIELDS[task]
    if for_path_only:
        return tuple(field for field in fields if field not in {"resolution", "house", "subtitle_type"}) + (FIELD_SKU,)
    return fields + (FIELD_SKU,)


def normalize_path_language(value: str) -> str:
    language = value.strip()
    if language not in LANGUAGE_OPTIONS:
        raise ValueError("Language must be English or Spanish.")
    return language


def normalize_sku(value: str) -> str:
    sku = value.strip().lower()
    if not sku:
        raise ValueError("SKU / Parent SKU is required.")
    if not re.fullmatch(r"[a-z0-9][a-z0-9-]*", sku):
        raise ValueError("SKU / Parent SKU can only use letters, numbers, and hyphens.")
    return sku


def normalize_season_folder(value: str) -> str:
    raw = value.strip().lower()
    if raw.startswith("s"):
        raw = raw[1:]
    if not raw.isdigit():
        raise ValueError("Season must be a number like 02.")
    return f"s{raw.zfill(2)}"


def normalize_year_season_folder(value: str) -> str:
    raw = value.strip().lower()
    if raw.startswith("s"):
        raw = raw[1:]
    if not re.fullmatch(r"\d{4}", raw):
        raise ValueError("Year must be four digits.")
    return f"s{raw}"


def normalize_episode_folder(value: str) -> str:
    raw = value.strip().lower()
    if raw.startswith("e"):
        raw = raw[1:]
    if not raw.isdigit():
        raise ValueError("Episode must be a number like 08.")
    return f"e{raw.zfill(2)}"


def title_slug(raw_fields: dict[str, str], label: str = "Title") -> str:
    slug = slugify(raw_fields.get(FIELD_TITLE, ""))
    if not slug:
        raise ValueError(f"{label} is required.")
    return slug


def localized_feature_folder(language: str) -> str:
    return "feature_las" if language == "Spanish" else "feature"


def localized_season_folder(season_folder: str, language: str) -> str:
    return f"{season_folder}_las" if language == "Spanish" else season_folder


def root_for_task(task: str) -> str:
    if task in MOVIE_ROOT_TASKS:
        return "movies"
    if task in SERIES_ROOT_TASKS:
        return "series"
    raise ValueError("Unsupported PathMaker task type.")


def project_slug_for_task(task: str, raw_fields: dict[str, str]) -> str:
    if task == "Exclusive Conversation (Yearly)":
        return "exclusive_conversation"
    label = "Series Title" if task == PATHMAKER_SERIES_TASK or "Episode" in task else "Title"
    return title_slug(raw_fields, label)


def project_folder(root: str, slug: str, sku: str) -> str:
    return f"s3://{PATHMAKER_BUCKET}/{root}/{slug}_{sku}"


def extra_folder(raw_fields: dict[str, str]) -> str:
    usage = raw_fields.get(FIELD_EXTRA_USAGE, "").strip()
    prefix = PATHMAKER_EXTRA_USAGE_TO_PREFIX.get(usage)
    if not prefix:
        raise ValueError("Choose an Extra Type supported by the source file structure doc.")
    usage_slug = slugify(usage)
    if not usage_slug:
        raise ValueError("Extra Type is required.")
    return f"{prefix}_{usage_slug}"


def build_pathmaker_path(task: str, raw_fields: dict[str, str]) -> str:
    if task not in PATHMAKER_TASK_FIELDS:
        raise ValueError("Unsupported PathMaker task type.")

    language = normalize_path_language(raw_fields.get(FIELD_LANGUAGE, PATHMAKER_DEFAULTS[FIELD_LANGUAGE]))
    sku = normalize_sku(raw_fields.get(FIELD_SKU, ""))
    root = root_for_task(task)
    project_slug = project_slug_for_task(task, raw_fields)
    base = project_folder(root, project_slug, sku)

    if task == PATHMAKER_SERIES_TASK:
        return f"{base}/"

    if task in {"Movie", "Caption", "Dub Audio"}:
        return f"{base}/{localized_feature_folder(language)}/"

    if task in {"Episode", "Episode Caption"}:
        season = localized_season_folder(normalize_season_folder(raw_fields.get(FIELD_SEASON, "")), language)
        episode = normalize_episode_folder(raw_fields.get(FIELD_EPISODE, ""))
        return f"{base}/{season}/{episode}/"

    if task in {"Original Premium Series (Yearly)", "Exclusive Conversation (Yearly)"}:
        season = localized_season_folder(normalize_year_season_folder(raw_fields.get(FIELD_YEAR, "")), language)
        episode = normalize_episode_folder(raw_fields.get(FIELD_EPISODE, ""))
        return f"{base}/{season}/{episode}/"

    if task == "Virtual Screening":
        return f"{base}/feature_virtual_screening/"

    if task in {"Virtual Screening Episode", "Virtual Screening Episode Caption"}:
        season = localized_season_folder(normalize_season_folder(raw_fields.get(FIELD_SEASON, "")), language)
        episode = normalize_episode_folder(raw_fields.get(FIELD_EPISODE, ""))
        return f"{base}/{season}/{episode}_virtual_screening/"

    if task in {"Trailer", "Trailer Caption"}:
        return f"{base}/trailer/"

    if task == "Extras":
        return f"{base}/extras/{extra_folder(raw_fields)}/"

    raise ValueError("Unsupported PathMaker task type.")
